chain/chain/src/resharding/trie_state_resharder.rs

```diff
@@ -278,3 +289,194 @@ impl TrieStateResharderMetrics {
+        // The resharding status should be None after completion.
+        assert!(resharder.load_status().unwrap().is_none());
+        check_child_tries_contain_all_keys(&test);
+    }
+
+    #[test]
+    fn test_trie_state_resharder_interrupt_and_resume() {
+        let test = setup_test();
+
+        let config = ChainConfig::test().resharding_config;
+        let resharder =
+            TrieStateResharder::new(test.runtime.clone(), ReshardingHandle::new(), config);
+
+        // Set the batch size to 1, this should stop iteration after the first key.
+        resharder
+            .resharding_config
+            .update(ReshardingConfig { batch_size: ByteSize(1), ..ReshardingConfig::test() });
+        let mut update_status = test.as_status();
+        resharder.process_batch_and_update_status(&mut update_status).unwrap();
+
+        let got_status = resharder
+            .load_status()
+            .unwrap()
+            .expect("status should not be empty after processing one batch");
+        assert_eq!(update_status, got_status);
+        // The persisted status should indicate continuing from the expected next key,
+        // which is the first key in the left child.
+        assert_eq!(
+            &test.initial.first().unwrap().0,
+            update_status.left.as_ref().unwrap().next_key.as_ref().unwrap()
+        );
+
+        // Test cancelling the resharding operation.
+        resharder.handle.stop();
+        resharder.resharding_blocking_impl(&mut update_status).unwrap();
+        // The resharding status should not have changed after cancellation.
+        assert_eq!(got_status, update_status);
+
+        // Test resuming the resharding operation.
+        let config = ChainConfig::test().resharding_config;
+        let resharder =
+            TrieStateResharder::new(test.runtime.clone(), ReshardingHandle::new(), config);
+        resharder.resume(test.parent_shard).expect("resume should succeed");
+
+        // The resharding status should be None after completion.
+        assert!(resharder.load_status().unwrap().is_none());
+        check_child_tries_contain_all_keys(&test);
+    }
+
+    fn check_child_tries_contain_all_keys(test: &TestSetup) {
+        let tries = test.runtime.get_tries();
+        // Using view_trie to bypass the memtrie and read from disk.
+        let left_trie = tries.get_view_trie_for_shard(test.left_shard, test.left_root);
+        let left_kvs = left_trie.lock_for_iter().iter().unwrap().map(Result::unwrap).collect_vec();
+        let right_trie = tries.get_view_trie_for_shard(test.right_shard, test.right_root);
+        let right_kvs =
+            right_trie.lock_for_iter().iter().unwrap().map(Result::unwrap).collect_vec();
+
+        // Iterate the child shards and check we can access the key/values.
+        let expected_keys = &test.initial;
+        expected_keys.iter().zip_eq(left_kvs.iter().chain(right_kvs.iter())).for_each(
```

Comment:
That works, because `left_kvs.iter().chain(right_kvs.iter())` are already sorted?

Reply:
yes. I will add a comment.